File: string/0678_valid_parenthesis_string/0678_valid_parenthesis_string.py
class Solution(object):
    def checkValidString(self, s):
        lower, upper = 0, 0
        for c in s:
            lower += 1 if c == '(' else -1
            upper -= 1 if c == ')' else -1
            if upper <0 : break
            lower = max(lower, 0)
        return lower == 0 


s = "(*)"
res = Solution().checkValidString(s)
print(res)

File: string/0678_valid_parenthesis_string/0678_valid_parenthesis_string.go
package main 

import "fmt"

func checkValidString(s string) bool {
	l, r := 0, 0
	n := len(s)
	for i:= 0;i<n;i++ {
		if s[i] == ')' {
			l--
		} else {
			l++
		}

		j := n-i-1 
		if s[j] == '(' {
			r--
		} else {
			r++
		}

		if l < 0 || r < 0 {
			return false
		}
	}

	return true
}

func main() {
	s := "(*)"
	res := checkValidString(s)
	fmt.Println(res)
}